internal/vpmem/types.go

```diff
@@ -0,0 +1,122 @@
+package vpmem
```

Comment:
I still feel like there's a lottttt of complexity in this file/implementation. I want to think some more about how we can reduce that if at all possible.